### PR TITLE
RST-399 interactive map rotation

### PR DESCRIPTION
**Pull request by svwilliams**
Created an rviz tool to allow the karto maps to be aligned. The tool creates an interactive marker to help align the maps generated by slam_karto. Two alignment endpoints are created with an arrow between them. The user can move the endpoints to align the arrow with a horizontal edge in the map. When the user clicks 'Align Map', a transform is publishd such that the arrow drawn in the 'map' frame will become aligned with the X-axis of the 'aligned' frame. slam_karto has been modified to check for this transform and render the map in the 'aligned' frame instead of the 'map' frame. Thus, this is not simply a rotation of the map; the map is regenerated in a different frame. This prevents aliasing and interpolation errors that would occur if the image was simply rotated. Note that map generation only occurs every 5 seconds, so there may be some lag between clicking "Align Map" and the aligned map being displayed.

Video: https://locusrobotics.slack.com/files/swilliams/F5XT5EUBA/map_alignment2.mp4

https://locusrobotics.atlassian.net/browse/RST-375

### CODE REVIEW

File: src/map_alignment_tool.cpp [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  }
+}
+
+void MapAlignmentTool::alignMapCallback(const visualization_msgs::InteractiveMarkerFeedbackConstPtr& feedback)
+{
+  // Query the two alignment handle markers from the marker server
+  visualization_msgs::InteractiveMarker endpoint1;
+  visualization_msgs::InteractiveMarker endpoint2;
+  if (interactive_marker_server_.get(endpoint1_name, endpoint1)
+    && interactive_marker_server_.get(endpoint2_name, endpoint2))
```
Comment (tappan-at-git) [inline src/map_alignment_tool.cpp]: Style: should probably line up the && operator with the 'i' in 'interactive' on the first line. (Or move the && to the previous line and line up the i's.

File: src/map_alignment_tool.cpp [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    // Publish aligned->map frame transformation
+    geometry_msgs::TransformStamped aligned_to_map_transform;
+    aligned_to_map_transform.header.stamp = ros::Time(0, 0);
+    aligned_to_map_transform.header.frame_id = aligned_frame_;
+    aligned_to_map_transform.child_frame_id = map_frame_;
+    aligned_to_map_transform.transform.translation.x = 0;
+    aligned_to_map_transform.transform.translation.y = 0;
+    aligned_to_map_transform.transform.translation.z = 0;
+    tf2::Quaternion q;
+    q.setRPY(0, 0, -yaw);  // send the opposite rotation so the map frame will end up in the requested orientation
```
Comment (tappan-at-git) [inline src/map_alignment_tool.cpp]: Could use [tf::createQuaternionFromYaw](http://docs.ros.org/api/tf/html/c++/namespacetf.html#a9550bc7dce951899c5cb577f4d58b867). (Unless that causes problems with tf2?)

File: src/slam_karto.cpp [hunk 3]
```diff
@@ -920,8 +923,49 @@ SlamKarto::updateMap()
...
+    // Lookup the map->aligned transform
+    tf::StampedTransform transform;
+    tf_.lookupTransform(aligned_frame_, map_frame_, ros::Time(), transform);
+    double yaw = tf::getYaw(transform.getRotation());
+
+    // Make a deep copy of all karto laserscans, transforming the corrected pose into the aligned frame
+    karto::LocalizedRangeScanVector transformed_scans;
+    transformed_scans.reserve(scans.size());
+    karto::Transform map_to_aligned_transform(karto::Pose2(0, 0, yaw));
+    for (size_t i = 0; i < scans.size(); ++i)
```
Comment (tappan-at-git) [inline src/slam_karto.cpp]: How much does it cost to retransform the scans every update?

File: src/map_alignment_tool.cpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  // Read the frame names from the parameter server
+  private_node_handle.getParam("aligned_frame", aligned_frame_);
+  private_node_handle.getParam("map_frame", map_frame_);
+  // Create the map alignment visualization markers
+  visualization_msgs::InteractiveMarker endpoint1 = createEndpoint(endpoint1_name);
+  visualization_msgs::InteractiveMarker endpoint2 = createEndpoint(endpoint2_name);
+  visualization_msgs::InteractiveMarker connector = createConnector(connector_name);
+  // Set the default positions of the markers
+  endpoint1.pose.position.x = -1.0;
+  endpoint2.pose.position.x = +1.0;
```
Comment (svwilliams) [inline src/map_alignment_tool.cpp]: Are you complaining about the "+" or about the hard-coded initial conditions?

Comment (tappan-at-git) [inline src/map_alignment_tool.cpp]: The "+".

File: src/map_alignment_tool.cpp (hunk 2 above)
Comment (svwilliams) [inline src/map_alignment_tool.cpp]: It really annoys me that the tf2 namespace does not have an equivalent function.

Comment (tappan-at-git) [inline src/map_alignment_tool.cpp]: Strict upgrade my behind.

File: src/slam_karto.cpp (hunk 3 above)
Comment (svwilliams) [inline src/slam_karto.cpp]: For the recent quiet_barnum map, the copy-transform-delete sequence takes about 20ms at the end of the mission. The map construction takes >800ms at that same point.

Comment (tappan-at-git) [inline src/slam_karto.cpp]: OK, so noticeable but clearly not a major source of lag

Comment (svwilliams) [inline src/slam_karto.cpp]: Correct.

**Pull request merged**